zcash_client_sqlite/src/wallet/init.rs

```diff
@@ -19,19 +26,286 @@ use {
+                spent INTEGER,
+                FOREIGN KEY (tx) REFERENCES transactions(id_tx),
+                FOREIGN KEY (account) REFERENCES accounts(account),
+                FOREIGN KEY (spent) REFERENCES transactions(id_tx),
+                CONSTRAINT tx_output UNIQUE (tx, output_index)
+            );
+            CREATE TABLE IF NOT EXISTS sapling_witnesses (
+                id_witness INTEGER PRIMARY KEY,
+                note INTEGER NOT NULL,
+                block INTEGER NOT NULL,
+                witness BLOB NOT NULL,
+                FOREIGN KEY (note) REFERENCES received_notes(id_note),
+                FOREIGN KEY (block) REFERENCES blocks(height),
+                CONSTRAINT witness_height UNIQUE (note, block)
+            );
+            CREATE TABLE IF NOT EXISTS sent_notes (
+                id_note INTEGER PRIMARY KEY,
+                tx INTEGER NOT NULL,
+                output_index INTEGER NOT NULL,
+                from_account INTEGER NOT NULL,
+                address TEXT NOT NULL,
+                value INTEGER NOT NULL,
+                memo BLOB,
+                FOREIGN KEY (tx) REFERENCES transactions(id_tx),
+                FOREIGN KEY (from_account) REFERENCES accounts(account),
+                CONSTRAINT tx_output UNIQUE (tx, output_index)
+            );",
+        )?;
+        Ok(())
+    }
+
+    fn down(&self, _transaction: &Transaction) -> Result<(), SqliteClientError> {
+        // We should never down-migrate the first migration, as that can irreversibly
+        // destroy data.
+        panic!("Cannot revert the initial migration.");
+    }
+}
+
+struct WalletMigration1;
+
+migration!(
+    WalletMigration1,
+    "a2e0ed2e-8852-475e-b0a4-f154b15b9dbe",
+    ["bc4f5e57-d600-4b6c-990f-b3538f0bfce1"],
+    "Add support for receiving transparent UTXOs."
+);
+
+impl RusqliteMigration for WalletMigration1 {
+    type Error = SqliteClientError;
+
+    fn up(&self, transaction: &Transaction) -> Result<(), SqliteClientError> {
+        transaction.execute_batch(
+            "CREATE TABLE IF NOT EXISTS utxos (
+                id_utxo INTEGER PRIMARY KEY,
+                address TEXT NOT NULL,
+                prevout_txid BLOB NOT NULL,
+                prevout_idx INTEGER NOT NULL,
+                script BLOB NOT NULL,
+                value_zat INTEGER NOT NULL,
+                height INTEGER NOT NULL,
+                spent_in_tx INTEGER,
+                FOREIGN KEY (spent_in_tx) REFERENCES transactions(id_tx),
+                CONSTRAINT tx_outpoint UNIQUE (prevout_txid, prevout_idx)
+            );",
+        )?;
+        Ok(())
+    }
+
+    fn down(&self, transaction: &Transaction) -> Result<(), SqliteClientError> {
+        transaction.execute_batch("DROP TABLE utxos;")?;
+        Ok(())
+    }
+}
+
+struct WalletMigration2<P: consensus::Parameters> {
+    params: P,
+    seed: SecretVec<u8>,
+}
+
+impl<P: consensus::Parameters> Migration for WalletMigration2<P> {
```

Comment:
Why isn't this expressible using the `migration!` macro?

Reply:
It's because the macro doesn't deal properly with the `<P>` type parameter.